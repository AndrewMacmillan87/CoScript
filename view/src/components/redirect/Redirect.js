import React, { Fragment } from "react";
import { withRouter } from "react-router-dom";

import { stopSoundEffect } from "../../redux/auxiliary/typing-sound-effect";

// Redux
import { connect } from "react-redux";
import { createNewProject } from "../../redux/actions/documents";
import { logoutUser } from "../../redux/actions/auth";

import PropTypes from "prop-types";

import "./redirect.css";

export class Redirect extends React.Component {
    /**
     * Uses the the custom set 'from' property of the props
     * history object to decide where to redirect the user to
     * and what actions, if any, to perform while redirecting.
     * For example, when redirecting from the '/editor' route
     * to a newly created project url, the project is created
     * while the redirecting is taking place
     */
    async componentDidMount() {
        stopSoundEffect();
        const { history, location, createNewProject, logoutUser } = this.props;

        if (location.state && location.state.passwordChanged) {
            setTimeout(() => {
                history["from"] = "verify";
                this.props.history.push("/login");
            }, 4000);
        } else {
            switch (history.from) {
                case "editor":
                    await createNewProject(history.newProjectData);
                    const { projectData } = this.props;
                    setTimeout(() => {
                        this.props.history.push(`/editor${projectData.url}`);
                    }, 4000);
                    break;
                case "verify":
                    setTimeout(() => {
                        this.props.history.push("/login");
                    }, 4000);
                    break;
                case "publications":
                    setTimeout(() => {
                        this.props.history.push("/publications");
                    }, 4000);
                    break;
                case "logout":
                    await logoutUser();
                    setTimeout(() => {
                        this.props.history.push("/");
                    }, 4000);
                    break;
                default:
                    setTimeout(() => {
                        this.props.history.push("/");
                    }, 4000);
            }
        }
    }

    render() {
        const { history, location } = this.props;
        const spinner = require("../../img/loading.svg");

        let component;

        if (location.state && location.state.passwordChanged) {
            history["from"] = "verify";
        }

        switch (history.from) {
            case "delete":
                component = (
                    <Fragment>
                        <p className='verify-text'>Sorry to see you go :(</p>
                        <p className='verify-text'>Redirecting...</p>
                    </Fragment>
                );
                break;
            case "editor":
                component = (
                    <Fragment>
                        <p className='verify-text'>
                            Creating your new project...
                        </p>
                    </Fragment>
                );
                break;
            case "verify":
                component = (
                    <Fragment>
                        <p className='verify-text'>
                            Taking you back to login...
                        </p>
                    </Fragment>
                );
                break;
            case "publications":
                component = (
                    <Fragment>
                        <p className='verify-text'>
                            Publishing your project! :)
                        </p>
                    </Fragment>
                );
                break;
            default:
                component = (
                    <p className='verify-text'>
                        Logging you out. Redirecting...
                    </p>
                );
        }

        return (
            <main className='redirect-body'>
                <section className='redirect-container'>
                    <h1 className='coscript-h1'>CoScript</h1>
                    {component}
                    <div className='ajax-spinner-container'>
                        <img src={spinner} alt='loading spinner' />
                    </div>
                </section>
            </main>
        );
    }
}

Redirect.propTypes = {
    history: PropTypes.object,
    createNewProject: PropTypes.func.isRequired,
    logoutUser: PropTypes.func.isRequired,
};

const mapProps = (state) => ({
    projectData: state.documents.projectData,
});

export default connect(mapProps, { createNewProject, logoutUser })(
    withRouter(Redirect)
);
