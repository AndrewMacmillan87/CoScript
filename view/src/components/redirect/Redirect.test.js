import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import { Provider } from "react-redux";
import { reduxStore } from "../../redux/store/reduxStore";

import { Redirect } from "./Redirect";

describe("Redirect component", () => {
    let props, mockCreateNewProject, mockLogoutUser;
    beforeEach(() => {
        global.HTMLMediaElement.prototype.pause = jest.fn();
        global.scrollTo = jest.fn();
        mockCreateNewProject = jest.fn();
        mockLogoutUser = jest.fn();
        props = {
            history: {},
            createNewProject: mockCreateNewProject,
            logoutUser: mockLogoutUser,
            location: {
                state: {},
            },
        };
    });

    afterEach(() => {
        props.history.from = null;
    });

    test("renders without crashing", () => {
        const div = document.createElement("div");
        ReactDOM.render(
            <Provider store={reduxStore}>
                <Router>
                    <Redirect {...props} />
                </Router>
            </Provider>,
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });

    it("renders correctly", () => {
        const tree = renderer
            .create(
                <Provider store={reduxStore}>
                    <Router>
                        <Redirect {...props} />
                    </Router>
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render a main element with a className of redirect-body", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Redirect {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("main").length).toEqual(1);
        expect(wrapper.find({ className: "redirect-body" }).length).toEqual(1);
    });

    it("should render a section element with a className of redirect-container", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Redirect {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("section").length).toEqual(1);
        expect(
            wrapper.find({ className: "redirect-container" }).length
        ).toEqual(1);
    });

    it("should render an img element", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Redirect {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("img").length).toEqual(1);
    });

    it("should render the message 'Sorry to see you go :(' from 'delete'", () => {
        props.history.from = "delete";
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Redirect {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("p").at(0).instance().innerHTML).toEqual(
            "Sorry to see you go :("
        );
    });

    it("should render the message 'Creating your new project...' from 'editor'", () => {
        props.history.from = "editor";
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Redirect {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("p").at(0).instance().innerHTML).toEqual(
            "Creating your new project..."
        );
    });

    it("should render the message 'Taking you back to login...' from 'verify'", () => {
        props.history.from = "verify";
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Redirect {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("p").at(0).instance().innerHTML).toEqual(
            "Taking you back to login..."
        );
    });

    it("should render the message 'Publishing your project! :)' from 'publications'", () => {
        props.history.from = "publications";
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Redirect {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("p").at(0).instance().innerHTML).toEqual(
            "Publishing your project! :)"
        );
    });

    it("should render the message 'Logging you out. Redirecting...' by default", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Redirect {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("p").at(0).instance().innerHTML).toEqual(
            "Logging you out. Redirecting..."
        );
    });

    it("should call `createNewProject()` from 'editor'", async () => {
        props.history.from = "editor";
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Redirect {...props} />
                </Router>
            </Provider>
        );
        await expect(mockCreateNewProject).toBeCalled();
    });

    it("should call `logoutUser()` from 'logout'", async () => {
        props.history.from = "logout";
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Redirect {...props} />
                </Router>
            </Provider>
        );
        await expect(mockLogoutUser).toBeCalled();
    });
});
