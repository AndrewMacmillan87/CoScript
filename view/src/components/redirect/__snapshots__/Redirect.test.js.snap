// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redirect component renders correctly 1`] = `
<main
  className="redirect-body"
>
  <section
    className="redirect-container"
  >
    <h1
      className="coscript-h1"
    >
      CoScript
    </h1>
    <p
      className="verify-text"
    >
      Logging you out. Redirecting...
    </p>
    <div
      className="ajax-spinner-container"
    >
      <img
        alt="loading spinner"
        src={
          Object {
            "ReactComponent": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "__esModule": true,
            "default": "loading.svg",
          }
        }
      />
    </div>
  </section>
</main>
`;
